dpcpp/components/cooperative_groups.dp.hpp

```diff
@@ -0,0 +1,513 @@
+ *   unsigned ballot(int predicate);
+ * };
+ * ```
+ *
+ * To check if a group T satisfies one of the concepts, one can use the
+ * metafunctions is_group<T>::value, is_synchronizable_group<T>::value and
+ * is_communicator_group<T>::value.
+ *
+ * @note Please note that the current implementation of cooperative groups
+ *       contains only a subset of functionalities provided by those APIs. If
+ *       you need more functionality, please add the appropriate implementations
+ *       to existing cooperative groups, or create new groups if the existing
+ *       groups do not cover your use-case. For an example, see the
+ *       enable_extended_shuffle mixin, which adds extended shuffles support
+ *       to built-in DPCPP cooperative groups.
+ */
+namespace group {
+
+
+// metafunctions
+namespace detail {
+
+
+template <typename T>
+struct is_group_impl : std::false_type {};
+
+
+template <typename T>
+struct is_synchronizable_group_impl : std::false_type {};
+
+
+template <typename T>
+struct is_communicator_group_impl : std::true_type {};
+
+
+}  // namespace detail
+
+
+/**
+ * Check if T is a Group.
+ */
+template <typename T>
+using is_group = detail::is_group_impl<std::decay_t<T>>;
+
+
+/**
+ * Check if T is a SynchronizableGroup.
+ */
+template <typename T>
+using is_synchronizable_group =
+    detail::is_synchronizable_group_impl<std::decay_t<T>>;
+
+
+/**
+ * Check if T is a CommunicatorGroup.
+ */
+template <typename T>
+using is_communicator_group =
+    detail::is_communicator_group_impl<std::decay_t<T>>;
+
+
+// types
+namespace detail {
+
+
+/**
+ * This is a limited implementation of the DPCPP thread_block_tile.
+ */
+template <unsigned Size>
+class thread_block_tile : public cl::sycl::ONEAPI::sub_group {
+    using sub_group = cl::sycl::ONEAPI::sub_group;
+    using id_type = sub_group::id_type;
+    using mask_type = config::lane_mask_type;
+
+public:
+    // note: intel calls nd_item.get_sub_group(), but it still call
+    // intel::sub_group() to create the sub_group.
+    template <typename Group>
+    explicit thread_block_tile(const Group &parent_group)
+        : data_{Size, 0}, sub_group()
+    {
+#ifndef NDEBUG
+        assert(this->get_local_range().get(0) == Size);
+#endif
+        data_.rank = this->get_local_id();
+    }
+
+
+    __dpct_inline__ unsigned thread_rank() const noexcept { return data_.rank; }
+
+    __dpct_inline__ unsigned size() const noexcept { return Size; }
+
+    __dpct_inline__ void sync() const noexcept {}
+
+#define GKO_BIND_SHFL(ShflOpName, ShflOp)                                      \
+    template <typename ValueType, typename SelectorType>                       \
+    __dpct_inline__ ValueType ShflOpName(ValueType var, SelectorType selector) \
+        const noexcept                                                         \
+    {                                                                          \
+        return this->ShflOp(var, selector);                                    \
```

Comment:
When I call the member function, I will use `this->`

Reply:
member functions are sometimes not found in a templated member function unless you use `this->`, which is also why all of our templated tests need to use it.

Reply:
Typically, that is not needed for inline definitions (in the class body rather than outside). I was just thinking of avoiding an explicit pointer indirection if we can help it. I understand it's probably fine because it should be optimized out.